### ISSUE TITLE
Clarifications needed about Censorship Resistance claims

### ISSUE THREAD

**Issue by lucash-dev** [nostr-protocol/nostr#102]
This document makes very bold claims about how Nostr can be a censorship-resistant replacement for Twitter.

However, there are hardly any arguments about how that can be the case.

In particular the answer to "Users getting banned and servers being closed" presents only vague suggestions of how to proceed in case a user is banned, and why these strategies are expected to work.

A few example points that should be considered.

- **Users posting relay suggestions**: why wouldn't relays censor these messages, since they benefit only the competition, in particular if relaying notes is a paid service? Is there any mechanism for preventing these notes from being censored? Is there any mechanism to *detect* if such notes are censored?

- **Informing followers through "close friends" about migration to other relays**: if spreading word across other channels via the social graph is viable, why use Nostr in the first place? Why not just share messages with friends and let the content spread from them? This seems to imply that Nostr is only censorship-resistant if there's a secondary censorship-resistant means of communication that can be used as a backup. If that's the case, what's the value Nostr is adding?

- **Number of relays:** the document suggests there need only be a handful of relays for the protocol to be successful. Why is that the case? If there is only a handful of relays wouldn't these relays have to handle an enormous amount of traffic (in case the protocol becomes popular)? If running a relay is expensive, wouldn't it be a barrier for entry to new relays, thus making censorship more likely? If on the other hand there are hundreds or thousands of relays splitting traffic, doesn't it make Nostr little different than Mastodon? If there's a single user with 100M+ followers (like Twitter has), how would relays handle that level of traffic?

If there are answers for these questions, I think it would be useful to either add them to this document, or add links to further information. Without such answers (and I can think of several more questions) the case for Nostr is pretty weak.

**In general if there is a more extensive technical explanation on why this protocol is expected to have the claimed properties (maybe a paper or other publication), it would be good to have a link to it**. Please note I'm not talking about protocol specification, but proofs or informal arguments about the protocol's properties, in particular the claimed Censorship Resistance.

I understand it's a WIP and has only recently gained broader attention.

But if Nostr is to succeed it needs good explanations on why it might in fact achieve its stated goals.

**Comment by lucash-dev** [comment]
Please let me know if there's a better forum for asking those questions.

**Comment by fiatjaf** [comment]
This is a good place.

**Comment by cjackie** [comment]
> Informing followers through "close friends" about migration to other relays

NIP 35 is one mechanism to inform followers what relays to find the author. Another way is by redundancy. One author can post the set_metadata to multiple relays, and chances are client will eventually find it even say 90% of relays censors it. With that in mind, client should implement a "discovering mode" where it finds all updated relays for pubkeys.

> Number of relays: the document suggests there need only be a handful of relays for the protocol to be successful

I think nostr needs many relays to be successful in censorship resistance. But, clients need to be *super smart* to find relays for pubkeys, and be *efficient* to avoid being banned by relays. 

> why wouldn't relays censor these messages

They can. but the idea is there will be *some* relays for taking the messages, client will find these relays.

**Comment by lucash-dev** [comment]
I appreciate the explanations above. I will further study NIP-35. 

I find the other points above still vague though. For example, what makes you believe it can still work with 90% censorship rate? What does being *super smart* and *efficient* entail?

I think there should be specific argument identifying possible censorship attack scenarios, existing mechanisms to avoid them, and estimates of likelihood of success and cost.

**Comment by GeorgeTsagk** [comment]
> NIP 35 is one mechanism to inform followers what relays to find the author. Another way is by redundancy. One author can post the set_metadata to multiple relays, and chances are client will eventually find it even say 90% of relays censors it. With that in mind, client should implement a "discovering mode" where it finds all updated relays for pubkeys.

Being redundant (in terms of being connected to many relays) and executing "discovery" of relays that don't censor is very costly in resource usage and contributes to the non-scalability of nostr. Nostr is very simple and that is good, but lacks structure in many ways (like any relation between content and where to search for it).

> I think nostr needs many relays to be successful in censorship resistance. But, clients need to be super smart to find relays for pubkeys, and be efficient to avoid being banned by relays.

Another factor that adds to the difficulty of this issue is the fact that you are not aware whether you are being served censored content or not. This again (as stated by @lucash-dev [here](https://github.com/nostr-protocol/nostr/issues/102#issue-1509817733)) calls for a secondary censorship resistant medium to address the problem.

> I find the other points above still vague though. For example, what makes you believe it can still work with 90% censorship rate? 

It can always work with 90% censorship rate but in theory. If you always connect to the right relay at the right time your content can always remain uncensored, but this is a great feat of discoverability.

> What does being super smart and efficient entail?

This is one of the challenges to the above issue. At high censorship rates you always have the chance of discovering the relays that don't censor but that comes with a great cost to the network. The lack of structure or relation of content/location adds to this problem as clients will have to randomly hit relays that they are aware of hoping to find the "hole" they're seeking.

**Comment by lucash-dev** [comment]
> It can always work with 90% censorship rate but in theory. If you always connect to the right relay at the right time your content can always remain uncensored, but this is a great feat of discoverability.

One problem with a "redundant servers" network compared to a peer-to-peer one, is that you need any two users to be connected to *at least one common, honest server* in order to exchange messages. In a peer-to-peer network all that would be required is a *path of honest nodes in the graph* between the two users. Either way, if you can't know if your peers are honest it is still a problem, but it is at least more likely you'll connect indirectly to the other users.

One way to address this would be to make relays also connect to each other and propagate messages to each other. That essentially turns it into a p2p network of servers, with "light clients" connecting to any subset of servers, which is probably too much of a departure from the architecture.

**Comment by GeorgeTsagk** [comment]
> One way to address this would be to make relays also connect to each other and propagate messages to each other.

I believe this is crucial for scalability. Say nostr has 100M users and 1M relays in the future. How much redundancy is there going to be needed for all of those 100M users to be able to retrieve content from each other through these 1M relays, without any filtering or censoring? You can always be partially connected to the "social graph" of relays according to your contacts & interests, but for that to work as intended there needs to be external communication about which relays to use.

> That essentially turns it into a p2p network of servers

I think this is the only possible way this could go. The main page of nostr states that it "does not rely on p2p techniques, therefore it works" which is a subtle argument that p2p techniques are bad/useless..?

P2P networks that were built in the past lacked one thing and that was a healthy incentive system for peers to participate. It is misleading to claim that not relying on p2p techniques is the reason nostr works. The fact that nostr is not a p2p network is the main reason it won't scale, as not treating relays as "peers" blocks you from re-using their content & effort in providing the service.

> which is probably too much of a departure from the architecture.

It is my personal opinion that it is worth re-structuring the base architecture as early as possible. Nostr has too much traction and it works right now, but we should make sure it will keep working in the future under different conditions.

**Comment by fiatjaf** [comment]
See more here: https://github.com/nostr-protocol/nips/issues/121

**Comment by fiatjaf** [comment]
@GeorgeTsagk yes, it's one of Nostr's core principles that P2P networks do not work, and yes, lack of incentive is probably the main reason why they don't work. Do you have an idea of how to fix these incentive problems now? Because you're basically suggesting that Nostr implements an overlay p2p network on top of relays, but I don't see the specifics of your proposal and how it would be better than past p2p networks.

If you want to experiment with a p2p network that is compatible with the current existing Nostr that would be a very welcome addition. If the keys and event formats are the same than we could easily have hybrid clients that talk to the two networks and be friendly about that. If the p2p network proves to be very reliable we can slowly transition everything to it -- although I think there will always be a place for the non-p2p part of the stack, but I don't know. I also would like to see p2p working as it is so elegant.

**Comment by lucash-dev** [comment]
> @GeorgeTsagk yes, it's one of Nostr's core principles that P2P networks do not work, and yes, lack of incentive is probably the main reason why they don't work. Do you have an idea of how to fix these incentive problems now? Because you're basically suggesting that Nostr implements an overlay p2p network on top of relays, but I don't see the specifics of your proposal and how it would be better than past p2p networks.
> 
> If you want to experiment with a p2p network that is compatible with the current existing Nostr that would be a very welcome addition. If the keys and event formats are the same than we could easily have hybrid clients that talk to the two networks and be friendly about that. If the p2p network proves to be very reliable we can slowly transition everything to it -- although I think there will always be a place for the non-p2p part of the stack, but I don't know. I also would like to see p2p working as it is so elegant.

I agree creating a separate project that is Nostr-compatible (or is its own thing completely) is the right approach if one wants to add a p2p layer. Like I said above it's too much of a departure from the basic architecture.

**Comment by lucash-dev** [comment]
> Sorry, @lucash-dev, I couldn't answer you earlier, but I've elaborated more on these points in my answers to this issue: [nostr-protocol/nips#121](https://github.com/nostr-protocol/nips/issues/121)

Thank you for the reference to the other discussion thread.

From all the bits of info in different places, what I can see that Nostr implements a number of adhoc countermeasures for the scenario where a user is banned or is otherwise censored.

However a general framework for *detecting the censorship* and exactly how to use the countermeasures is missing.

As for detecting censorship, I think there are two easy steps that can be taken on a user level that would go a long way. They can be automated but not necessarily need to be part of the core protocol:

- Include in each post a sequence number.
- Agree on a minimum rate of posts per day or week.

If followers see missing sequence numbers, they know some posts have been either censored or lost, and would know they should look for other relays to try and find the content, thus preventing undetected censorship of individual notes (this could include relay hints).

The second strategy would make censoring all notes starting from a time easily detectable.

Of course the user posting the messages could mess up with the sequences or just stop posting -- so this isn't proof of censorship. Presumably followers (or their software) can decide how much resources is it worth spending on trying to find the missing posts before giving up.

The author of the posts could try and make anonymized queries to the relays to test them for censorship, in case the relay serves back uncensored posts to the author but not to other users.

(maybe this was already suggested or implemented elsewhere?)

**Comment by GeorgeTsagk** [comment]
> @GeorgeTsagk yes, it's one of Nostr's core principles that P2P networks do not work, and yes, lack of incentive is probably the main reason why they don't work. Do you have an idea of how to fix these incentive problems now? Because you're basically suggesting that Nostr implements an overlay p2p network on top of relays, but I don't see the specifics of your proposal and how it would be better than past p2p networks.

I do not aim in suggesting a specific proposal for a p2p network over relays. That would transform nostr's design to a more complicated one, and I believe the simplicity of nostr is something to hold on to. Like @lucash-dev mentioned, this would be too much of a departure from the basic architecture.

I think it's all about how the documentation is placed.

On the `README.md`: "_it does not rely on P2P techniques, therefore it works._"

I just believe the above phrase is inaccurate.

Also within the same README:

> - Censorship-resistance
>    - Each user can publish their updates to any number of relays.
>    - A relay can charge a fee (the negotiation of that fee is outside of the protocol for now) from users to publish there, which ensures censorship-resistance (there will always be some Russian server willing to take your money in exchange for serving your posts).

None of the above 2 sub-bullets prove anything about censorship resistance.

I believe that, like this issue's title suggests, some of the claims in the main documentation page need rephrasing.

**Comment by lucash-dev** [comment]
> On the `README.md`: "_it does not rely on P2P techniques, therefore it works._"
> 
> I just believe the above phrase is inaccurate.

100% agreed.

**Comment by fiatjaf** [comment]
The readme of course is not trying to "prove anything". But how would someone go about "proving" censorship-resistance?

**Comment by lucash-dev** [comment]
> The readme of course is not trying to "prove anything". But how would someone go about "proving" censorship-resistance?

I agree the readme doesn't need to prove anything. But bold claims should be backed by proofs or strong arguments that can be found *somewhere*.

As for proving censorship resistance, there's quite some scientific literature on the subject. Though there isn't a single generally accepted formal definition, there are a few related definitions, and some proofs of censorship resistance of protocols under different sets of assumptions. 

I haven't seen much of a formal proof of Bitcoin's censorship resistance -- but the argument (quite obviously true), first made in the original paper, is that an attacker can't censor transactions (prevent them from being included in the consensus chain) as long as a considerable majority of the hash rate belongs to honest miners (possibly more than 51% due to selfish mining attacks), and there's at least a path of reachable honest nodes between the originator of a transaction and a honest miner. There's also a strong economic incentive for miners to be honest. 

A couple discussions of censorship resistance and proofs relating to other kinds of protocols:

- https://www.freehaven.net/anonbib/cache/ih05-csispir.pdf
- https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=5ca963f25befa66a80a2511734e65c2b8128c9f5
- https://www.freehaven.net/anonbib/cache/Serj02-iptps.pdf

Please note that censorship-resistance in the context of public pseudonymous messages is not going to be the same as in private point-to-point anonymous communications, but that might give an idea.

I am working on some related research, and have some (unpublished) tentative definitions -- though my goals aren't exactly the same as Nostr's. If I find a definition that fits what Nostr seems to be attempting to achieve, and any proofs that it is (or isn't) censorship-resistant I'll let you know.

**Comment by dontfwiththestream** [comment]
the simple fact that relays can censor whatever content they want means this will end up a splintered mess like mastodon. Censorship is literally built into the software, calling it censorship resistant when mechanisms exist to censor is disingenuous at best

**Comment by zzo38** [comment]
A key signing party might be able to verify that keys are correct, and might also be able to check if any messages are missing that should not be missing. Even outside of a key signing party, people who meet in person might be able to verify if something is wrong.

Also, relays can be recommended inside of tags in other messages, and since these tags are signed, they cannot easily be tampered with.

Sequence numbers is one way to check that messages are not lost, but I think that a minimum rate is not always desirable (and sometimes might not work, e.g. in case you are sick and cannot use the computer, or if the computer is broken, a server is temporarily down for repairs, etc).

Allowing relays to communicate with each other might also help to avoid some messages getting lost (whether deliberately or not).

One other consideration is for messages to contain links to older messages, to be "blockchaining". However, there are problems with this too due to sometimes such links will not work, messages can get lost by accident, etc. But it can be an optional feature if desired. If it is desirable, you can link to both your own messages and to other people's messages, so that you can "cross-blockchain" them, to make the links "stronger" and then you can detect if any message is missing.

**Comment by futureisfoss** [comment]
I also agree that its misleading to say "it does not rely on p2p techniques, therefore it works" on the main page of Nostr. For example, one p2p network I do know that works well is BitTorrent. Speaking of that, I think it'd be cool to have something like Nostr that also has BitTorrent support. The sweet spot for me is when a network relies on servers for seamless transfer of data but also isn't completely dependent on those servers because of p2p support. Right now if a post becomes very popular on Nostr then the relays of that user will have to bear the load, now imagine if those relays were seeding the post using BitTorrent then other users could also help with the sharing. But one issue with BitTorrent is that it might leak a user's IP address, so it should be either opt-in or used in conjunction with a VPN or I2P network to hide user's IP. Because this thread is about censorship resistance, I also wanna point out that using BitTorrent over i2p would be almost impossible to censor, albeit your transfer speeds would be very slow.

I know that Nostr probably won't add these, but I just wanted to throw the idea out there in case someone wants to make something like that.

**Comment by Pantyhose-X** [comment]
Remove child pornography videos and images is a must.
I can't accept the Remove of NSFL, horror, [murder](https://sos.photos/), [violence](https://sos.photos/), videos and images, because these are real occurrences, so that more people can condemn them, so that people know that life is precious. 99% of the news is covering up these truths.

**Comment by marcelolucario** [comment]
> This censorship is very common when certain content violates human rights, duties or obligations.

I think censorship is done according to the government, the laws. In this sense, I imagine a denunciation channel made up of 5 participants: 
- "1. general users, end-users, home-users, ", 
- "2. developers"
- "3. government"
- "4. non-governmental entities"
- "5. companies or business". 

So, this can be done with the reporting channel observed with certain confidence metrics:
> 1. Complaint between users to/between users
> 2. Complaint between governments to/between governments
> 4. Complaint between companies to/between companies
> 5. Complaint between non-governmental entities to/between non-governmental entities
> 6. Complaint between non-governmental entities to/between governments
> 7. Complaint between non-governmental entities to/between governments
> 8. Complaint between politicians entities to/between governments
> 9. Complaint between workers entities to/between governments
> ... etc

What I'm describing here would be the "invisible hand" of the market. The general idea is this - my idea is to analyze certain variables like:
> 1. whenever a user is reported, it is possible that one or more users will also be reported.
> 2. the only way for the user not to be denounced is if a certain number x of participants do not know or are conniving with such an attitude.
> 3. For something to be taken down, there must be x number of reports.
> 4. But there is a possibility that no participant wants to make the denunciation
> 5. But there is no possibility for any participant to want to make the denunciation
> 6. But there is a possibility that the content is legal. In this sense, the user is not reported.
> 7. Whenever a user is reported, it is necessary to communicate to the press, the government, the public prosecutor and lawyers.
> 8. Lawyers, the government, the press, the police receive the complaint and investigate the case.

1. For this theoretical scheme to work, it would be necessary for the user of the nostr network to enter his type of  function: public or private. Things like Discord, Discourse, Stackoverflow have the information of what role the person is: administrative, company manager, etc. This would be interesting, to address a transparent trust metric.
2. It would be necessary to have smart contracts to evaluate user conduct, complaints.
3. Artificial intelligence algorithms can recognize good and bad behavior patterns. At the same time, they can be allies to remove or report certain users with strange, inappropriate behavior.
4. In real life, it is not possible to exercise a profession without having some level of documentation or identification.

**Comment by milahu** [comment]
since you cannot force any relay to relay your posts, you would have to run your own relay

this is a fundamental limitation of the "relay" approach (nostr, email, twitter, IRC),
compared to a "tunnel" approach (tor hidden services), where both ends must be online at the same time

probably this is one reason why SSB (secure scuttlebutt) pulls the full chain of events (append-only log), because only then you can detect missing events

[README.md#the-problem-with-ssb-secure-scuttlebutt](https://github.com/nostr-protocol/nostr#the-problem-with-ssb-secure-scuttlebutt)

> It insists on having a **chain of updates** from a single user, which feels unnecessary to me and something that adds bloat and rigidity to the thing — each server/user needs to store all the chain of posts to be sure the new one is valid. Why? (Maybe they have a good reason);
